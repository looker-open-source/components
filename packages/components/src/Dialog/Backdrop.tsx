/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import {
  CompatibleHTMLProps,
  reset,
  shouldForwardProp,
} from '@looker/design-tokens'
import { OpacityProps, BackgroundColorProps, color } from 'styled-system'
import styled, { CSSObject } from 'styled-components'

export interface BackdropProps
  extends CompatibleHTMLProps<HTMLDivElement>,
    BackgroundColorProps,
    OpacityProps {
  visible?: boolean
  inlineStyle?: CSSObject
}

// Backdrop styles are applied here (rather than using the inline `style={...}` prop) to ensure that
// transitions will still apply to backdrop
export const Backdrop = styled.div
  .withConfig({ shouldForwardProp })
  .attrs(() => ({
    'data-testid': 'backdrop',
  }))`
  ${reset}
  ${color}

  background: ${({ theme }) => theme.colors.ui5};
  bottom: 0;
  cursor: default;
  left: 0;
  opacity: 0.6;
  position: fixed;
  right: 0;
  top: 0;
  transition: opacity ${({ theme }) => theme.transitions.simple}ms;

  &.entering,
  &.exiting {
    opacity: 0.01;
  }
`
