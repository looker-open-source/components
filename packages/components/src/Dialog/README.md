# Dialog

Dialogs break out of the standard application flow and UI to present new information or required actions.

Dialog provides a general purpose (empty & unstyled) overlay. `Confirm` will likely be more useful if your intent is to render a standard user confirmation dialog.

## Standard Use

`Dialog` operates in a "uncontrolled" manner.

```tsx
export function Basic(props: DialogProps) {
  return (
    <Dialog {...props} content="Simple Content">
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

```tsx
export function MediumContent(props: DialogProps) {
  const DialogExampleLayout = () => {
    const { closeModal } = React.useContext(DialogContext);

    return (
      <DialogLayout
        header="Simple header"
        footer={
          <>
            <Button onClick={closeModal}>Done Reading</Button>
            <ButtonTransparent color="neutral" onClick={closeModal}>
              Finish Later
            </ButtonTransparent>
          </>
        }
      >
        Lorem Ipsum is simply dummy text of the printing and typesetting
        industry. Lorem Ipsum has been the industry\' s standard dummy text ever
        since the 1500s, when an unknown printer took a galley of type and
        scrambled it to make a type specimen book.
      </DialogLayout>
    );
  };

  return (
    <Dialog {...props} content={<DialogExampleLayout />}>
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

## Controlled

`Dialog` can be used in a "controlled" manner by specifying the `isOpen` property. An `onClose` callback should also be specified so that clicking the dialog's backdrop or calling the `DialogContext.closeModal` method will work as expected.

```tsx
export function Controlled(props: DialogProps) {
  const [isOpen, setIsOpen] = React.useState(false);

  return (
    <>
      <Dialog
        {...props}
        isOpen={isOpen}
        onClose={() => setIsOpen(false)}
        content={
          <DialogLayout
            header="Simple header"
            footer={
              <>
                <Button onClick={() => setIsOpen(false)}>Done Reading</Button>
                <ButtonTransparent
                  color="neutral"
                  onClick={() => setIsOpen(false)}
                >
                  Finish Later
                </ButtonTransparent>
              </>
            }
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting
            industry. Lorem Ipsum has been the industry\' s standard dummy text
            ever since the 1500s, when an unknown printer took a galley of type
            and scrambled it to make a type specimen book.
          </DialogLayout>
        }
      />
      <Button onClick={() => setIsOpen(true)}>Open Dialog</Button>
    </>
  );
}
```

## Hook: useDialog

`Dialog` is also offered as a React Hook via `useDialog`. It returns the `open` callback and rendered dialog.

```tsx
export function UseDialog(props: DialogProps) {
  const { dialog, setOpen } = useDialog({
    ...props,
    content: 'My Neat Dialog',
  });

  return (
    <>
      {dialog}
      <ButtonOutline onClick={() => setOpen(true)}>Open Dialog</ButtonOutline>
    </>
  );
}

export function WithCheckbox() {
  return (
    <Dialog
      content={
        <DialogLayout footer="Footer" header="Header">
          The top line & bottom shadow should not be there.
          <Checkbox checked />
        </DialogLayout>
      }
    >
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

## onAfterClose and onAfterOpen function

onAfterClose: this function will be called after the `Dialog` is closed
onAfterOpne: this function will be called after the `Dialog` is open

```tsx
export function AnimationCallbacks(props: DialogProps) {
  return (
    <Dialog
      {...props}
      content="Simple Content"
      onAfterClose={() => {
        alert('Close');
      }}
      onAfterOpen={() => {
        alert('Open');
      }}
    >
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

## Terminology

`Dialogs` are assembled of two pieces: an overlay and a surface. The `Backdrop` component is the semi-opaque full-screen overlay which signals that the the rest of the application is inaccessible. The `Surface` component sits on top of the overlay and renders the relevant content.

The most common pattern for a `Dialog` is the [`Confirm`](/components/dialogs/confirm) pattern.

### Backdrop

This provides the backdrop behind dialogs.

### Surface

`Surface` provides the container that contains the content.

## Width & minHeight

All variants of Dialog allow you to specify `width` - it can be specified as `small`, `medium`, & `large`.

By default, `width` defaults to `medium` so that Dialog Surface will conform to the width of its content. At the same time, `maxWidth` constrains the Dialog surface's width to be no larger than the specified value.

Specify a `minHeight` to prevent a `Dialog` from getting any shorter than the specified value.

```tsx
export function Height(props: DialogProps) {
  const { closeModal } = React.useContext(DialogContext);

  return (
    <Dialog
      {...props}
      height="1000rem"
      content={
        <DialogLayout
          header="Simple header"
          footer={
            <>
              <Button onClick={closeModal}>Done Reading</Button>
              <ButtonTransparent color="neutral" onClick={closeModal}>
                Finish Later
              </ButtonTransparent>
            </>
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting
          industry. Lorem Ipsum has been the industry\' s standard dummy text
          ever since the 1500s, when an unknown printer took a galley of type
          and scrambled it to make a type specimen book.
        </DialogLayout>
      }
    >
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

## Placement

Determines how surface is positioned in the the viewport.

- `center` (default) - surface is centered horizontally and vertically above mobile breakpoint.
  - _mobile_: positioned at top of window and covers entire width.
  - _width_: defaults to 100% in mobile breakpoint and 37.5rem above that unless otherwise specified
  - _height_: fits content unless it's explicitly specified with `minHeight` prop
- `cover` - positioned to cover nearly the entire window.
  - _mobile & tablet_: cover the entire window.

```tsx
export function PlacementCover(props: DialogProps) {
  const { closeModal } = React.useContext(DialogContext);

  return (
    <Dialog
      {...props}
      placement="cover"
      content={
        <DialogLayout
          header="Simple header"
          footer={
            <>
              <Button onClick={closeModal}>Done Reading</Button>
              <ButtonTransparent color="neutral" onClick={closeModal}>
                Finish Later
              </ButtonTransparent>
            </>
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting
          industry. Lorem Ipsum has been the industry\' s standard dummy text
          ever since the 1500s, when an unknown printer took a galley of type
          and scrambled it to make a type specimen book.
        </DialogLayout>
      }
    >
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

- `top` - vertically positioned near the top of edge of the window, horizontally centered.
  - _mobile_: identical to `default` placement
  - _height_: grows to fit content. Total height will keep surface a small amount from top and bottom of viewport
  - _width_: default `medium` above mobile breakpoint

```tsx
export function PlacementTop(props: DialogProps) {
  const { closeModal } = React.useContext(DialogContext);

  return (
    <Dialog
      {...props}
      placement="top"
      content={
        <DialogLayout
          header="Simple header"
          footer={
            <>
              <Button onClick={closeModal}>Done Reading</Button>
              <ButtonTransparent color="neutral" onClick={closeModal}>
                Finish Later
              </ButtonTransparent>
            </>
          }
        >
          Lorem Ipsum is simply dummy text of the printing and typesetting
          industry. Lorem Ipsum has been the industry\' s standard dummy text
          ever since the 1500s, when an unknown printer took a galley of type
          and scrambled it to make a type specimen book.
        </DialogLayout>
      }
    >
      <ButtonOutline>Open Dialog</ButtonOutline>
    </Dialog>
  );
}
```

## DialogContext

```tsx
import React, { useContext } from 'react';
import { Button, DialogContext } from '@looker/components';

export function DialogContextDemo() {
  return (
    <DialogContext.Consumer>
      {({ close }) => <Button onClick={close}>Close!</Button>}
    </DialogContext.Consumer>
  );
}

export function DailogUseContextDemo() {
  const { close } = useContext(DialogContext);
  return <Button onClick={close}>Close!</Button>;
}
```

`DialogContext` is a [React Context](https://reactjs.org/docs/context.html) that provides
access to the `closeModal` function, which allows you to close the Dialog or Popover.

## Scroll Lock

A key part of dialog usability is the scroll lock, which disables scrolling on any part of the page except within the dialog.
When the `Dialog` component renders, it registers its portal element to both the `ScrollLockContext` and `FocusTrapContext`
in `ComponentsProvider`, which manage competing scroll locks and focus traps from other components.

In rare cases, such as nesting a popover from another library inside a `Dialog`, you may need to temporarily disable these behaviors:

```tsx
import { ScrollLockContext } from '@looker/components-providers';
import { useContext } from 'react';

const { activeTrapRef, disableCurrentTrap, enableCurrentTrap } =
  useContext(ScrollLockContext);

function toggleScrollLock() {
  if (activeTrapRef && activeTrapRef.current) {
    disableCurrentTrap();
  } else {
    enableCurrentTrap();
  }
}
```
