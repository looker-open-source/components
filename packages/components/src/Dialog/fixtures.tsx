/*

 MIT License

 Copyright (c) 2024 Google LLC

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { useState } from 'react';
import { Button, ButtonTransparent } from '../Button';
import { Dialog, DialogLayout } from '.';
import type { DialogProps } from '.';

export function Controlled(props: DialogProps) {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <>
      <Dialog
        {...props}
        isOpen={isOpen}
        onClose={() => setIsOpen(false)}
        content={
          <DialogLayout
            header="Simple header"
            footer={
              <>
                <Button onClick={() => setIsOpen(false)}>Done Reading</Button>
                <ButtonTransparent
                  color="neutral"
                  onClick={() => setIsOpen(false)}
                >
                  Finish Later
                </ButtonTransparent>
              </>
            }
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting
            industry. Lorem Ipsum has been the industry\' s standard dummy text
            ever since the 1500s, when an unknown printer took a galley of type
            and scrambled it to make a type specimen book.
          </DialogLayout>
        }
      />
      <Button onClick={() => setIsOpen(true)}>Open Dialog</Button>
    </>
  );
}

export function ControlledLegacy(props: DialogProps) {
  const [isOpen, setIsOpen] = React.useState(false);

  return (
    <>
      <Dialog
        {...props}
        isOpen={isOpen}
        onClose={() => setIsOpen(false)}
        content={
          <DialogLayout
            header="Simple header"
            footer={
              <>
                <Button onClick={() => setIsOpen(false)}>Done Reading</Button>
                <ButtonTransparent
                  color="neutral"
                  onClick={() => setIsOpen(false)}
                >
                  Finish Later
                </ButtonTransparent>
              </>
            }
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting
            industry. Lorem Ipsum has been the industry\' s standard dummy text
            ever since the 1500s, when an unknown printer took a galley of type
            and scrambled it to make a type specimen book.
          </DialogLayout>
        }
      />
      <button onClick={() => setIsOpen(true)}>Open Dialog</button>
    </>
  );
}

export function ControlledNoChildren(props: DialogProps) {
  const [isOpen, setIsOpen] = React.useState(false);

  return (
    <>
      <Dialog
        {...props}
        isOpen={isOpen}
        onClose={() => setIsOpen(false)}
        content={
          <DialogLayout
            header="Simple header"
            footer={
              <>
                <Button onClick={() => setIsOpen(false)}>Done Reading</Button>
                <ButtonTransparent
                  color="neutral"
                  onClick={() => setIsOpen(false)}
                >
                  Finish Later
                </ButtonTransparent>
              </>
            }
          >
            Lorem Ipsum is simply dummy text of the printing and typesetting
            industry. Lorem Ipsum has been the industry\' s standard dummy text
            ever since the 1500s, when an unknown printer took a galley of type
            and scrambled it to make a type specimen book.
          </DialogLayout>
        }
      />
      <button onClick={() => setIsOpen(true)}>Open Dialog</button>
    </>
  );
}

export function CloseIconButton(props: DialogProps) {
  return (
    <Dialog
      {...props}
      content={
        <DialogLayout header="Has a close icon button">
          Some content
        </DialogLayout>
      }
    >
      <Button>Open Dialog</Button>
    </Dialog>
  );
}
