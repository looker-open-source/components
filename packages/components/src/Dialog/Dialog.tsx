/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React, { CSSProperties, FC } from 'react'
import { CSSTransition } from 'react-transition-group'
import { CSSObject } from 'styled-components'
import { ResponsiveValue } from 'styled-system'
import { Portal } from '../Portal'
import { useFocusTrap, useScrollLock } from '../utils'
import { Backdrop } from './Backdrop'
import { DialogContext } from './DialogContext'
import { Surface } from './Surface'

export interface DialogProps {
  /**
   * When true, renders the Backdrop, Surface and it's contained content.
   * @default false
   */
  isOpen?: boolean

  /**
   * Specify a callback to be called each time this Dialog is closed
   */
  onClose?: () => void

  /**
   * Optional backdrop styles to merge with the Backdrop implementation. These
   * must be a CSSProperty compatible key / value paired object. For example
   * {backgroundColor: 'pink'}.
   */
  backdrop?: CSSProperties

  /**
   * Optional surface styles to merge with the Surface implementation. These
   * must be a CSSProperty compatible key / value paired object.
   */
  surfaceStyles?: CSSProperties

  /**
   * Explicitly specifying a width will set the Surface to be the lesser of the specified width or the viewport width.
   * You can also specify `auto` if you want the Surface to auto-size to its content.
   * @default auto
   */
  width?: ResponsiveValue<string>
  maxWidth?: ResponsiveValue<string>
}

export const Dialog: FC<DialogProps> = ({
  backdrop,
  children,
  isOpen,
  onClose,
  maxWidth,
  surfaceStyles,
  width,
}) => {
  const {
    callbackRef: focusRef,
    disable: disableFocusTrap,
    enable: enableFocusTrap,
    isEnabled: focusTrapEnabled,
    trapRef: focusTrapRef,
  } = useFocusTrap(isOpen)
  const {
    callbackRef: scrollRef,
    disable: disableScrollLock,
    enable: enableScrollLock,
    isEnabled: scrollLockEnabled,
  } = useScrollLock(isOpen, false)

  const handleClose = () => {
    onClose && onClose()
  }

  return (
    <DialogContext.Provider
      value={{
        closeModal: handleClose,
        disableFocusTrap,
        disableScrollLock,
        enableFocusTrap,
        enableScrollLock,
        focusTrapEnabled,
        focusTrapRef,
        scrollLockEnabled,
      }}
    >
      <CSSTransition
        classNames="modal"
        mountOnEnter
        unmountOnExit
        in={isOpen}
        timeout={{ enter: 0, exit: 250 }}
      >
        {(state: string) => (
          <Portal
            ref={(node) => {
              focusRef(node)
              scrollRef(node)
            }}
          >
            <Backdrop
              className={state}
              onClick={onClose}
              visible={backdrop === undefined ? true : !!backdrop}
              style={
                !!backdrop && backdrop !== true
                  ? (backdrop as CSSObject)
                  : undefined
              }
            />
            <Surface
              style={surfaceStyles}
              className={state}
              width={width}
              maxWidth={maxWidth}
            >
              {children}
            </Surface>
          </Portal>
        )}
      </CSSTransition>
    </DialogContext.Provider>
  )
}
