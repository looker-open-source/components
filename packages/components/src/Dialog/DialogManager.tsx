/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import omit from 'lodash/omit'
import React, {
  cloneElement,
  FC,
  isValidElement,
  ReactNode,
  useState,
} from 'react'
import { Dialog, DialogProps } from './Dialog'

type DialogRenderProp = (dialogProps: {
  onClick: () => void
  className?: string
  role?: string
  'aria-expanded'?: boolean
}) => ReactNode

function isRenderProp(
  children: ReactNode | DialogRenderProp
): children is DialogRenderProp {
  return typeof children === 'function'
}

export interface DialogManagerProps extends DialogProps {
  children?: DialogRenderProp | ReactNode
  /**
   * Content that will be placed inside the Dialog
   * @required
   */
  content: ReactNode
  /**
   * Specify a callback to be called before trying to close the Dialog. This allows for
   * use-cases where the user might lose work (think common "Save before closing warning" type flow)
   * Specify a callback to be called each time this Dialog is closed
   */
  canClose?: () => boolean
}

export const DialogManager: FC<DialogManagerProps> = ({
  canClose,
  content,
  children,
  onClose,
  ...props
}) => {
  const [isOpen, setOpen] = useState(props.isOpen || false)
  const open = () => setOpen(true)
  const close = () => {
    if (canClose && !canClose()) return
    onClose && onClose()
    setOpen(false)
  }

  const dialogPropsLabeled = {
    'aria-expanded': isOpen,
    className: isOpen ? 'active' : '',
    onClick: open,
    role: 'button',
  }

  const dialog = (
    <Dialog isOpen={isOpen} onClose={close} {...omit(props, 'isOpen')}>
      {content}
    </Dialog>
  )

  let target = children

  if (isValidElement(children)) {
    target = cloneElement(children, {
      ...dialogPropsLabeled,
      className: isOpen
        ? `${children.props.className} active`
        : children.props.className,
    })
  } else if (isRenderProp(children)) {
    target = children(dialogPropsLabeled)
  } else {
    // eslint-disable-next-line no-console
    console.warn(
      `Element "${typeof target}" can't be used as target for DialogManager`
    )
  }

  return (
    <>
      {dialog}
      {target}
    </>
  )
}
