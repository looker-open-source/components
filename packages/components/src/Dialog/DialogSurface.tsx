/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import styled, { css } from 'styled-components'
import { minHeight, ResponsiveValue } from 'styled-system'
import { theme } from '@looker/design-tokens'
import { SurfaceBase, surfaceTransition } from '../Dialog/SurfaceBase'
import { dialogWidth, DialogWidth } from './dialogWidth'

export type DialogPlacements = 'center' | 'cover' | 'top'

export interface DialogSurfaceProps {
  /**
   * Determines how Surface is positioned in the the viewport.
   * `center` - surface is centered horizontally and vertically above mobile breakpoint.
   *    mobile: positioned at top of window and covers entire width.
   *    width: defaults to 100% in mobile breakpoint and 37.5rem above that unless otherwise specified
   *    height: fits content unless it's explicitly specified with `height` prop
   * `cover` - positioned to cover nearly the entire window.
   *    mobile & tablet: cover the entire window.
   * `top` - vertically positioned near the top of edge of the window, horizontally centered.
   *    mobile: identical to `default` placement
   *    height: grows to fit content. Total height will keep surface a small amount from top and bottom
   *      of viewport
   *    width: default `medium` above mobile breakpoint
   * @default 'center'
   */
  placement?: DialogPlacements

  /**
   * Explicitly specifying a width will set the Surface to be the lesser of
   * the specified width or the viewport width.
   */
  width?: DialogWidth

  /**
   * Explicitly specifying a minHeight will set the Surface to be the lesser of
   * the specified height, the content height or the viewport height. Default will
   * cause the Surface to auto-size to its content.
   */
  minHeight?: ResponsiveValue<string>
}

const { space, breakpoints } = theme
const gapSpace = 'xxlarge'
const coverDimension = `calc(100% - ${space[gapSpace]} * 2)`

const placements = {
  center: css<DialogSurfaceProps>`
    align-self: flex-start;
    max-height: 100%;

    @media screen and (min-width: ${breakpoints[0]}) {
      align-self: center;
      max-height: ${coverDimension};
    }
  `,
  cover: css<DialogSurfaceProps>`
    height: 100%;

    @media screen and (min-width: ${breakpoints[0]}) {
      height: ${coverDimension};
      width: ${coverDimension};
    }

    @media screen and (min-width: ${breakpoints[1]}) {
      height: ${coverDimension};
      width: ${coverDimension};
    }
  `,
  top: css<DialogSurfaceProps>`
    align-self: flex-start;
    margin-top: 0;
    max-height: 100%;

    @media screen and (min-width: ${breakpoints[0]}) {
      margin-top: ${({ theme }) => theme.space[gapSpace]};
      max-height: ${coverDimension};
    }
  `,
}

export const DialogSurface = styled(SurfaceBase)<DialogSurfaceProps>`
  box-shadow: ${({ theme }) => theme.shadows[5]};
  position: relative;
  transition: transform ${surfaceTransition}, opacity ${surfaceTransition};

  ${dialogWidth}
  ${({ placement }) => placements[placement || 'center']}
  ${minHeight}

  @media screen and (min-width: ${breakpoints[0]}) {
    border-radius: ${({ theme }) => theme.radii.medium};
  }

  &.entering,
  &.exiting {
    opacity: 0.01;
    transform: translateY(100%);
  }
`

DialogSurface.defaultProps = {
  width: 'medium',
}
