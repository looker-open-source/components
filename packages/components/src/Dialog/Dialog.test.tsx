/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import 'jest-styled-components'
import React from 'react'
import {
  assertSnapshotShallow,
  mountWithTheme,
} from '@looker/components-test-utils'
import { Portal } from '../Portal'
import { Backdrop } from './Backdrop'
import { Dialog } from './Dialog'
import { DialogManager } from './DialogManager'

const simpleContent = (
  <div>
    simple content
    <button>Done</button>
  </div>
)

describe('Dialog', () => {
  test('Active', () => {
    assertSnapshotShallow(<Dialog isOpen>{simpleContent}</Dialog>)
  })

  test('Verify initial state', () => {
    const dialog = mountWithTheme(<Dialog>{simpleContent}</Dialog>)

    expect(dialog.find(Portal).exists()).toEqual(false)
    expect(dialog.contains(simpleContent)).toBeFalsy()
  })

  test('Verify "open" prop', () => {
    const dialog = mountWithTheme(<Dialog isOpen>{simpleContent}</Dialog>)

    expect(dialog.find(Portal).exists()).toEqual(true)
    expect(dialog.contains(simpleContent)).toBeTruthy()
  })

  describe('Dialog styled', () => {
    const dialog = mountWithTheme(
      <Dialog
        isOpen
        backdrop={{ backgroundColor: 'pink' }}
        surfaceStyles={{ backgroundColor: 'purple' }}
      >
        {simpleContent}
      </Dialog>
    )

    test('Dialog applies the backdrop styles', () => {
      const backdrop = dialog.find(Backdrop)

      expect(backdrop.exists()).toBeTruthy()

      expect(backdrop.props().style).toEqual({ backgroundColor: 'pink' })
    })
  })
})

describe('DialogManager - click events', () => {
  test('Trigger.click renders a backdrop, clicking backdrop closes it', () => {
    const dialog = mountWithTheme(
      <DialogManager content={simpleContent}>
        {({ onClick }) => <a onClick={onClick}>Open Dialog</a>}
      </DialogManager>
    )

    // Dialog closed
    expect(dialog.find(Portal).exists()).toBeFalsy()

    const button = dialog.find('a')
    expect(button.exists()).toBeTruthy()
    button.simulate('click') // Click to open

    // Dialog open
    expect(dialog.find(Portal).exists()).toBeTruthy()

    const backdrop = dialog.find(Backdrop)
    expect(backdrop.exists()).toBeTruthy()
    window.document.body.click()
  })
})

test('contains the content passed to it', () => {
  const dialog = mountWithTheme(
    <DialogManager content={simpleContent}>
      <a>Open Dialog</a>
    </DialogManager>
  )

  const button = dialog.find('a')
  expect(button.exists()).toBeTruthy()
  button.simulate('click') // Click to open
  expect(dialog.contains(simpleContent)).toBeTruthy()
  window.document.body.click()
})
