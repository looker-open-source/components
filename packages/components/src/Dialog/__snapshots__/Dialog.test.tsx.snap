// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog Active 1`] = `ShallowWrapper {}`;
